Add new argument tags to existing Inception projects

We would like to be able to add new sets of tags to existing projects without having to add them manually (which would be extremely tedious depending on how many projects must be modified). For example, an ontology extension gives all `Life.Die(.subtype)` event types a `Killer` argument role, so all applicable projects will need a `Killer` tag added if it doesn't have one already.

Since the Inception application doesn't appear to allow mass tag addition, I thought of this plan which uses the remote API:
1. **Export and save** all relevant project files (e.g. all `Life.Die` projects).
2. Create modified copies of these ZIP archives such that for each `.json` project file, **add the new tags** for that event type (e.g. add `Killer` to `tag_set[tags]`).
3. We want to import these modified projects without creating a bunch of `copy_of` projects, which is what Inception automatically does to prevent projects from being overwritten (and this doesn't appear to be configurable), so **delete** each project from Inception that was successfully modified locally.
4. For each modified project, **import** these to Inception.

@gabbard : The most relevant thing I can find is this:
```
05:30:05.763 [http-nio-8080-exec-1371] ERROR de.tudarmstadt.ukp.clarin.webanno.w
ebapp.remoteapi.aero.AeroRemoteApiController - Channel not open for writing - ca
nnot extend file to required size
java.io.IOException: Channel not open for writing - cannot extend file to requir
ed size
        at sun.nio.ch.FileChannelImpl.map(FileChannelImpl.java:898) ~[?:1.8.0_22
2]
        at sun.nio.ch.FileChannelImpl.transferFromFileChannel(FileChannelImpl.ja
va:631) ~[?:1.8.0_222]
        at sun.nio.ch.FileChannelImpl.transferFrom(FileChannelImpl.java:702) ~[?
:1.8.0_222]
        at org.apache.commons.io.FileUtils.doCopyFile(FileUtils.java:1131) ~[com
mons-io-2.6.jar!/:2.6]
        at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1076) ~[commo
ns-io-2.6.jar!/:2.6]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:10
00) ~[commons-io-2.6.jar!/:2.6]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:96
1) ~[commons-io-2.6.jar!/:2.6]
        at de.tudarmstadt.ukp.clarin.webanno.api.dao.export.exporters.Annotation
DocumentExporter.exportAnnotationDocumentContents(AnnotationDocumentExporter.jav
a:249) ~[webanno-api-dao-4.0.0-beta-12.jar!/:?]
        at de.tudarmstadt.ukp.clarin.webanno.api.dao.export.exporters.Annotation
DocumentExporter.exportData(AnnotationDocumentExporter.java:123) ~[webanno-api-d
ao-4.0.0-beta-12.jar!/:?]
        at de.tudarmstadt.ukp.clarin.webanno.api.dao.export.ProjectExportService
Impl.exportProject(ProjectExportServiceImpl.java:222) ~[webanno-api-dao-4.0.0-be
ta-12.jar!/:?]
...
```
but it seems more likely to be from the export errors that occurred last night.

Here's an extended traceback, if it's helpful:
```
14:42:55.531 [main] WARN  mu.KLogging - com.github.kittinunf.fuel.core.BubbleFuelError: HTTP Exception 400 
	com.github.kittinunf.fuel.core.FuelError$Companion.wrap(FuelError.kt:84)
	com.github.kittinunf.fuel.core.DeserializableKt.response(Deserializable.kt:168)
	edu.isi.vista.annotationutils.TestAddNewTagsKt$main$$inlined$use$lambda$1.invoke(TestAddNewTags.kt:323)
	edu.isi.vista.annotationutils.TestAddNewTagsKt$main$$inlined$use$lambda$1.invoke(TestAddNewTags.kt)
	edu.isi.vista.annotationutils.ExportAnnotationsKt.retryOnFuelError(ExportAnnotations.kt:259)
	edu.isi.vista.annotationutils.ExportAnnotationsKt.retryOnFuelError$default(ExportAnnotations.kt:256)
	edu.isi.vista.annotationutils.TestAddNewTagsKt.main(TestAddNewTags.kt:187)
Caused by: com.github.kittinunf.fuel.core.BubbleFuelError: HTTP Exception 400
...
```

--- a/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
+++ b/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
@@ -121,13 +121,14 @@ fun main(argv: Array<String>) {
     val projectsToEventTypes = mutableMapOf<Project, String>()
     for (serverProject in serverProjects) {
         val projectName = serverProject.name
-        var abridgedProjectName = serverProject.name
         // Ignore any language prefixes so that it knows
         // to modify *all* projects of this event type
-        for (language in languagePrefixes) {
-            if (projectName.startsWith(language)) {
-                abridgedProjectName = abridgedProjectName.removePrefix("$language-")
-            }
+        val abridgedProjectName = if (languagePrefixes.any {projectName.startsWith(it)}) {
+            val firstHyphen = projectName.indexOf('-')
+            val language = projectName.substring(0, firstHyphen)
+            projectName.removePrefix("$language-")
+        } else {
+            projectName
         }
         for (eventType in eventTypesToModify) {
             // If the project's name minus language prefix begins with the project name to edit:
@@ -223,7 +224,6 @@ fun main(argv: Array<String>) {
                 // Step 4: import the new projects.
                 logger.info { "Importing project $modifiedZipOutputFile" }
                 retryOnFuelError {
-                    // The documentation is a lie.
                     importUrl.httpUpload()
                             .add(
                                     FileDataPart(

--- a/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
+++ b/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
@@ -1,0 +1,202 @@
+package edu.isi.vista.annotationutils
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.github.kittinunf.fuel.core.Request
+import com.github.kittinunf.fuel.core.extensions.authentication
+import com.github.kittinunf.fuel.httpDelete
+import com.github.kittinunf.fuel.httpGet
+import com.github.kittinunf.fuel.httpPost
+import edu.isi.nlp.parameters.serifstyle.SerifStyleParameterFileLoader
+import net.java.truevfs.comp.zip.ZipFile
+import java.io.File
+import java.net.URI
+import java.nio.file.FileSystems
+import java.nio.file.Files
+
+/*
+Takes existing Inception projects and adds new argument labels.
+
+This is useful if there are ontology changes to adopt.
+
+This will add tags only to AIDA event type projects by default.
+To add tags to event types of other projects, add the project
+prefix to the event types list. (e.g. "ACE-Conflict.Attack")
+
+A new tag can be added to all projects of a specific type by simply
+making the event type the project prefix (e.g. "ACE": ["NewArgument"])
+*/
+
+fun main(argv: Array<String>)  {
+    if (argv.size != 1) {
+        throw RuntimeException("Expected a single argument, a parameter file")
+    }
+    val paramsLoader = SerifStyleParameterFileLoader.Builder()
+            .interpolateEnvironmentalVariables(true).build()
+    val params = paramsLoader.load(File(argv[0]))
+
+    // load parameters
+    val inceptionUrl = params.getString("inceptionUrl")
+    val inceptionUserName = params.getString("inceptionUsername")
+    val inceptionPassword = params.getString("inceptionPassword")
+
+    val projectDir = params.getExistingDirectory("project_dir")
+    val usersList = params.getOptionalExistingFile("users_list")
+    val modifiedOutputDir = params.getCreatableDirectory("modified_output_dir")
+    // One may want add arguments for specific users
+    if (usersList != null) {
+        val users = usersList.orNull()?.readLines()?.toSet()
+    }
+
+//    val outputDirectory = params.getCreatableDirectory("output_dir")!!
+//    outputDirectory.mkdirs()
+
+    // handles JSON (de)serialization
+    val objectMapper = ObjectMapper()
+    val mapper = ObjectMapper().registerKotlinModule()
+    val prettyPrinter = objectMapper.writerWithDefaultPrettyPrinter()
+
+    // Load the relevant event types and their new tags
+    val eventTypesToNewTags = objectMapper.readTree(params.getExistingFile("new_tag_list")) as ObjectNode
+    val eventTypesToModify: MutableSet<String> = eventTypesToNewTags.fieldNames().asSequence().toMutableSet()
+    val languagePrefixes = arrayListOf("russian", "spanish")
+
+    // For each project listed in the new tags list:
+    for (eventType in eventTypesToModify) {
+        val arguments = eventTypesToNewTags[eventType]
+        // For each project in project_dir:
+        projectDir.walk().filter { it.isFile }.forEach { zipFile ->
+            val filename = zipFile.name
+            var abridgedFilename = zipFile.name
+            // Ignore any language prefixes
+            for (language in languagePrefixes) {
+                if (filename.startsWith(language)) {
+                    abridgedFilename = abridgedFilename.removePrefix("$language-")
+                }
+            }
+            // If the project's name minus language prefix begins with the project name to edit:
+            if (abridgedFilename.startsWith(eventType)) {
+                logger.info { "$filename is about to be modified." }
+
+                ZipFile(zipFile.toPath()).use {
+                    val jsonBytes = it.getInputStream("exportedproject123.json")?.readBytes()
+                    if (jsonBytes != null) {
+                        val jsonTree = objectMapper.readTree(jsonBytes)
+
+                        // tag sets - these will be added to below for each argument
+                        val tagSet = (jsonTree["tag_sets"] as ArrayNode)[0] as ObjectNode
+                        val tags = tagSet["tags"] as ArrayNode
+                        logger.info { "Tags: $tags" }
+                        // first layer is for spans, second for argument relation type
+                        val argumentTypeLayer = (jsonTree["layers"] as ArrayNode)[1]
+                        val argumentTypeTags = (argumentTypeLayer["features"] as ArrayNode)[0]["tag_set"]["tags"] as ArrayNode
+
+                        // Add arguments
+                        for (argument in arguments) {
+                            val tagJson = objectMapper.createObjectNode()
+                            tagJson.put("tag_name", argument)
+                            tagJson["tag_description"] = null
+                            if (tags.contains(argument))
+                                logger.info { "$filename already has tag $argument!" }
+                            tags.add(tagJson)
+                            argumentTypeTags.add(tagJson)
+                        }
+
+                        val zipOutFile = modifiedOutputDir.resolve(filename)
+                        // we need to delete any existing file or else the zip file system will just add
+                        // to its content
+                        zipOutFile.delete()
+                        val env = mapOf("create" to "true")
+                        val uri = URI.create("jar:file:$zipOutFile")
+
+                        FileSystems.newFileSystem(uri, env).use { zipfs ->
+                            val pathInZipfile = zipfs.getPath("exportedproject123.json")
+                            Files.newBufferedWriter(pathInZipfile).use { out ->
+                                out.write(prettyPrinter.writeValueAsString(jsonTree))
+                            }
+                        }
+                    } else {
+                        throw RuntimeException("Corrupt zip file returned")
+                    }
+                }
+            }
+        }
+    }
+    // Now that we have our new projects, we import them to Inception.
+    // Since we don't want severl "copy_of" projects, we delete the
+    // original project files as long as the projects had no
+    // extension function to avoid authentication boilerplate
+    fun Request.authenticateToInception() =
+            this.authentication().basic(inceptionUserName, inceptionPassword)
+
+    val projects = retryOnFuelError {
+        "$inceptionUrl/api/aero/v1/projects".httpGet()
+                .authenticateToInception()
+                .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
+                .body
+    }
+            ?: throw java.lang.RuntimeException("Could not fetch projects from $inceptionUrl. Aborting")
+    logger.info { "Projects on server ${projects.map { it.name }}" }
+
+    val canBeDeleted = arrayListOf<String>()
+    val projectsToDelete = arrayListOf<String>()
+    val projectsToImport = arrayListOf<File>()
+    for (project in projects) {
+        logger.info { "Processing project $project" }
+        val projectName = project.name
+        var abridgedProjectName = project.name
+        for (language in languagePrefixes) {
+            if (project.name.startsWith(language)) {
+                abridgedProjectName = abridgedProjectName.removePrefix("$language-")
+            }
+        }
+        val getDocsUrl = "$inceptionUrl/api/aero/v1/projects/${project.id}/documents"
+        val documents = retryOnFuelError {
+            getDocsUrl.httpGet(
+                    parameters = listOf("projectId" to project.id))
+                    .authenticateToInception()
+                    .resultObjectThrowingExceptionOnFailure<DocumentResult>(mapper)
+                    .body
+        }
+        if (documents == null) {
+            logger.warn { "Skipping $project due to web errors" }
+            continue
+        }
+        // Projects without annotated documents can be
+        if (documents.isEmpty() and eventTypesToModify.any {abridgedProjectName.startsWith(it)}) {
+            // logger.info { "$project has no documents"}
+            canBeDeleted.add(projectName)
+        }
+    }
+    logger.info { "Projects that can safely be deleted: $canBeDeleted" }  // testing
+    modifiedOutputDir.walk().filter { it.isFile }.forEach { zipFile ->
+        val projectName = zipFile.nameWithoutExtension
+        if (canBeDeleted.contains(projectName)) {
+            projectsToDelete.add(projectName)
+            projectsToImport.add(zipFile)
+        }
+    }
+    logger.info { "Projects to replace: $projectsToImport" }
+    // Delete projects with replacements
+//    for (project in canBeDeleted) {
+//        logger.info { "Deleting project $project" }
+//        retryOnFuelError {
+//            "$inceptionUrl/api/aero/v1/projects/$project".httpDelete()
+//                    .authenticateToInception()
+//                    .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
+//                    .body
+//        }
+//    }
+//    // Import updated projects to replace deleted ones
+//    for (newProject in projectsToImport) {
+//        logger.info { "Importing project $newProject" }
+//        retryOnFuelError {
+//            "$inceptionUrl/api/aero/v1/projects/import/$newProject".httpPost()
+//                    .authenticateToInception()
+//                    .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
+//                    .body
+//        }
+//    }
+}

--- a/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
+++ b/annotation-utils/src/main/java/edu/isi/vista/annotationutils/AddNewTags.kt
@@ -1,35 +1,76 @@
 package edu.isi.vista.annotationutils
 
+import com.fasterxml.jackson.databind.JsonNode
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.databind.node.ArrayNode
 import com.fasterxml.jackson.databind.node.ObjectNode
 import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.github.kittinunf.fuel.core.FileDataPart
 import com.github.kittinunf.fuel.core.Request
 import com.github.kittinunf.fuel.core.extensions.authentication
 import com.github.kittinunf.fuel.httpDelete
 import com.github.kittinunf.fuel.httpGet
-import com.github.kittinunf.fuel.httpPost
+import com.github.kittinunf.fuel.httpUpload
+import com.google.common.jimfs.Configuration
+import com.google.common.jimfs.Jimfs
 import edu.isi.nlp.parameters.serifstyle.SerifStyleParameterFileLoader
+import net.java.truevfs.comp.zip.ZipEntry
 import net.java.truevfs.comp.zip.ZipFile
+import net.java.truevfs.comp.zip.ZipOutputStream
+import java.io.BufferedOutputStream
 import java.io.File
-import java.net.URI
-import java.nio.file.FileSystems
 import java.nio.file.Files
 
-/*
-Takes existing Inception projects and adds new argument labels.
+/**
+ * Takes existing Inception projects and adds new argument labels.
+ *
+ * This is useful if there are ontology changes to adopt.
+ *
+ * This will add tags only to AIDA event type projects (i.e. those not
+ * prefixed by something other than a language) by default.
+ * To add tags to event types of other projects, add the project
+ * prefix to the event types list. (e.g. "ACE-Conflict.Attack")
+ * Note that a language prefix is not needed to add tags to
+ * non-English projects.
+ *
+ * A new tag can be added to all projects of a specific type by simply
+ * making the event type the project prefix (e.g. "ACE": ["NewArgument"])
+ *
+ * It may be a good idea to run this with the parameters
+ * ```
+ * modifyLocalOnly: true
+ * originalExportDir: <path_to_separate_backup_directory>
+ * ```
+ * first to be extra-sure that you have backups to (re-)import manually
+ * in case something goes awry.
+ *
+ * Required parameters:
+ * <ul>
+ *     <li> `inceptionUrl` is the base URL of your Inception server.</li>
+ *     <li> `inceptionUserName` is the user name of an Inception user with access to the
+ *          remote API. This privilege must be granted in the Inception administration interface.
+ *          </li>
+ *     <li> `inceptionPassword` is the password for the user above. Note that parameter files can
+ *          interpolate from environmental variables</li>
+ *     <li> `originalExportDir` is the directory where the original project ZIP files
+ *     will be saved. These are handy in case any projects need to be re-imported.</li>
+ *     <li> `modifiedOutputDir` is the directory where the projects with added tags
+ *     will be saved.</li>
+ *     <li> `userList` is the text file containing the list of users whose projects
+ *     will be modified.</li>
+ *     <li> `newTagList` is the JSON file containing the event types and their new tags.
+ *     (An example of how this should look is in `docs/aida_changes_0514.json`.)</li>
+ * </ul>
+ *
+ * Optional parameters:
+ * <ul>
+ *     <li> `modifyLocalOnly`: if true, none of the modified projects will be imported
+ *     to Inception, and their original projects will not be deleted from the server.
+ *     If false (default), the program will run its normal course.</li>
+ * </ul>
+ */
 
-This is useful if there are ontology changes to adopt.
-
-This will add tags only to AIDA event type projects by default.
-To add tags to event types of other projects, add the project
-prefix to the event types list. (e.g. "ACE-Conflict.Attack")
-
-A new tag can be added to all projects of a specific type by simply
-making the event type the project prefix (e.g. "ACE": ["NewArgument"])
-*/
-
-fun main(argv: Array<String>)  {
+fun main(argv: Array<String>) {
     if (argv.size != 1) {
         throw RuntimeException("Expected a single argument, a parameter file")
     }
@@ -42,161 +83,202 @@ fun main(argv: Array<String>)  {
     val inceptionUserName = params.getString("inceptionUsername")
     val inceptionPassword = params.getString("inceptionPassword")
 
-    val projectDir = params.getExistingDirectory("project_dir")
-    val usersList = params.getOptionalExistingFile("users_list")
-    val modifiedOutputDir = params.getCreatableDirectory("modified_output_dir")
-    // One may want add arguments for specific users
-    if (usersList != null) {
-        val users = usersList.orNull()?.readLines()?.toSet()
-    }
-
-//    val outputDirectory = params.getCreatableDirectory("output_dir")!!
-//    outputDirectory.mkdirs()
+    val originalExportDir = params.getCreatableDirectory("originalExportDir")
+    val userSet = params.getExistingFile("userList").readLines().toSet()
+    val modifiedOutputDir = params.getCreatableDirectory("modifiedOutputDir")
+    val modifyLocalOnly = params.getOptionalBoolean("modifyLocalOnly").or(false)
 
     // handles JSON (de)serialization
     val objectMapper = ObjectMapper()
     val mapper = ObjectMapper().registerKotlinModule()
     val prettyPrinter = objectMapper.writerWithDefaultPrettyPrinter()
 
+    // Part I: Create new project files with new tags
+
     // Load the relevant event types and their new tags
-    val eventTypesToNewTags = objectMapper.readTree(params.getExistingFile("new_tag_list")) as ObjectNode
+    val newTagListFile = params.getExistingFile("newTagList")
+    val eventTypesToNewTags = objectMapper.readTree(newTagListFile) as ObjectNode
     val eventTypesToModify: MutableSet<String> = eventTypesToNewTags.fieldNames().asSequence().toMutableSet()
     val languagePrefixes = arrayListOf("russian", "spanish")
 
-    // For each project listed in the new tags list:
-    for (eventType in eventTypesToModify) {
-        val arguments = eventTypesToNewTags[eventType]
-        // For each project in project_dir:
-        projectDir.walk().filter { it.isFile }.forEach { zipFile ->
-            val filename = zipFile.name
-            var abridgedFilename = zipFile.name
-            // Ignore any language prefixes
-            for (language in languagePrefixes) {
-                if (filename.startsWith(language)) {
-                    abridgedFilename = abridgedFilename.removePrefix("$language-")
-                }
-            }
-            // If the project's name minus language prefix begins with the project name to edit:
-            if (abridgedFilename.startsWith(eventType)) {
-                logger.info { "$filename is about to be modified." }
-
-                ZipFile(zipFile.toPath()).use {
-                    val jsonBytes = it.getInputStream("exportedproject123.json")?.readBytes()
-                    if (jsonBytes != null) {
-                        val jsonTree = objectMapper.readTree(jsonBytes)
-
-                        // tag sets - these will be added to below for each argument
-                        val tagSet = (jsonTree["tag_sets"] as ArrayNode)[0] as ObjectNode
-                        val tags = tagSet["tags"] as ArrayNode
-                        logger.info { "Tags: $tags" }
-                        // first layer is for spans, second for argument relation type
-                        val argumentTypeLayer = (jsonTree["layers"] as ArrayNode)[1]
-                        val argumentTypeTags = (argumentTypeLayer["features"] as ArrayNode)[0]["tag_set"]["tags"] as ArrayNode
-
-                        // Add arguments
-                        for (argument in arguments) {
-                            val tagJson = objectMapper.createObjectNode()
-                            tagJson.put("tag_name", argument)
-                            tagJson["tag_description"] = null
-                            if (tags.contains(argument))
-                                logger.info { "$filename already has tag $argument!" }
-                            tags.add(tagJson)
-                            argumentTypeTags.add(tagJson)
-                        }
-
-                        val zipOutFile = modifiedOutputDir.resolve(filename)
-                        // we need to delete any existing file or else the zip file system will just add
-                        // to its content
-                        zipOutFile.delete()
-                        val env = mapOf("create" to "true")
-                        val uri = URI.create("jar:file:$zipOutFile")
-
-                        FileSystems.newFileSystem(uri, env).use { zipfs ->
-                            val pathInZipfile = zipfs.getPath("exportedproject123.json")
-                            Files.newBufferedWriter(pathInZipfile).use { out ->
-                                out.write(prettyPrinter.writeValueAsString(jsonTree))
-                            }
-                        }
-                    } else {
-                        throw RuntimeException("Corrupt zip file returned")
-                    }
-                }
-            }
-        }
-    }
-    // Now that we have our new projects, we import them to Inception.
-    // Since we don't want severl "copy_of" projects, we delete the
-    // original project files as long as the projects had no
     // extension function to avoid authentication boilerplate
     fun Request.authenticateToInception() =
             this.authentication().basic(inceptionUserName, inceptionPassword)
 
-    val projects = retryOnFuelError {
+    // Get the list of projects on the server
+    val serverProjects = retryOnFuelError {
         "$inceptionUrl/api/aero/v1/projects".httpGet()
                 .authenticateToInception()
                 .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
                 .body
     }
             ?: throw java.lang.RuntimeException("Could not fetch projects from $inceptionUrl. Aborting")
-    logger.info { "Projects on server ${projects.map { it.name }}" }
+    logger.info { "Projects on server ${serverProjects.map { it.name }}" }
+    val importUrl = "$inceptionUrl/api/aero/v1/projects/import"
+    val failureSet = mutableSetOf<String>()  // will hold names of projects that were not modified
 
-    val canBeDeleted = arrayListOf<String>()
-    val projectsToDelete = arrayListOf<String>()
-    val projectsToImport = arrayListOf<File>()
-    for (project in projects) {
-        logger.info { "Processing project $project" }
-        val projectName = project.name
-        var abridgedProjectName = project.name
+    // Determine which projects need to be processed
+    val projectsToEventTypes = mutableMapOf<Project, String>()
+    for (serverProject in serverProjects) {
+        val projectName = serverProject.name
+        var abridgedProjectName = serverProject.name
+        // Ignore any language prefixes so that it knows
+        // to modify *all* projects of this event type
         for (language in languagePrefixes) {
-            if (project.name.startsWith(language)) {
+            if (projectName.startsWith(language)) {
                 abridgedProjectName = abridgedProjectName.removePrefix("$language-")
             }
         }
-        val getDocsUrl = "$inceptionUrl/api/aero/v1/projects/${project.id}/documents"
-        val documents = retryOnFuelError {
-            getDocsUrl.httpGet(
-                    parameters = listOf("projectId" to project.id))
+        for (eventType in eventTypesToModify) {
+            // If the project's name minus language prefix begins with the project name to edit:
+            if (abridgedProjectName.startsWith(eventType) and userSet.any { abridgedProjectName.contains(it) }) {
+                logger.info { "$projectName is about to be modified." }
+                projectsToEventTypes[serverProject] = eventType
+            }
+        }
+    }
+
+    for (project in projectsToEventTypes) {
+        // Step 1: export relevant projects from Inception.
+        logger.info { "Now exporting ${project.key.name}" }
+        val getExportUrl = "$inceptionUrl/api/aero/v1/projects/${project.key.id}/export.zip"
+        val exportZip = retryOnFuelError {
+            getExportUrl.httpGet()
                     .authenticateToInception()
-                    .resultObjectThrowingExceptionOnFailure<DocumentResult>(mapper)
-                    .body
+                    .retryOnResponseFailure()
         }
-        if (documents == null) {
-            logger.warn { "Skipping $project due to web errors" }
-            continue
-        }
-        // Projects without annotated documents can be
-        if (documents.isEmpty() and eventTypesToModify.any {abridgedProjectName.startsWith(it)}) {
-            // logger.info { "$project has no documents"}
-            canBeDeleted.add(projectName)
+        if (exportZip != null) {
+            // Step 2: add new tags by creating modified copies
+            // of the original projects.
+            val inMemoryFileSystem = Jimfs.newFileSystem(Configuration.unix())
+            val exportInMemoryZip = inMemoryFileSystem.getPath("/export.zip")
+            Files.write(exportInMemoryZip, exportZip)
+
+            // Save the zip file so that it can be re-imported later if needed
+            val zipOutputFile = File(originalExportDir, "${project.key.name}.zip")
+            val zipOutputPath = zipOutputFile.toPath()
+            Files.write(zipOutputPath, exportZip)
+            val modifiedZipOutputFile = modifiedOutputDir.resolve("${project.key.name}.zip")
+            val modifiedZipOutputPath = modifiedZipOutputFile.toPath()
+            val modifiedZipInMemory = inMemoryFileSystem.getPath("/modified_export.zip")
+            var replaceProject = false
+
+            ZipFile(exportInMemoryZip).use {
+                val zipOutputStream = ZipOutputStream(
+                        BufferedOutputStream(
+                                Files.newOutputStream(modifiedZipOutputPath)
+                        )
+                )
+                for (projectFile in it) {
+                    val inMemoryProjectPath = inMemoryFileSystem.getPath("/${projectFile.name}")
+                    val projectBytes = it.getInputStream(projectFile.name)?.readBytes()
+                    if (projectBytes != null) {
+                        if (inMemoryProjectPath.count() > 1) {
+                            // Create any parent directories if they don't already exist
+                            if (!Files.exists(inMemoryProjectPath.parent)) {
+                                Files.createDirectories(inMemoryProjectPath.parent)
+                            }
+                        }
+                        Files.write(inMemoryProjectPath, projectBytes)
+                    }
+                    if (projectFile.name.contains("exportedproject")) {
+                        val jsonBytes = it.getInputStream(projectFile.name)?.readBytes()
+                        if (jsonBytes != null) {
+                            Files.delete(inMemoryProjectPath)
+                            val inMemoryJson = inMemoryFileSystem.getPath("/${projectFile.name}")
+                            val jsonTree = objectMapper.readTree(jsonBytes)
+                            val arguments = eventTypesToNewTags[project.value]
+                            replaceProject = addArgumentsToJson(jsonTree, project.key.name, arguments)
+                            val jsonToWrite = prettyPrinter.writeValueAsString(jsonTree)
+                            Files.write(inMemoryJson, jsonToWrite.toByteArray())
+                            zipOutputStream.putNextEntry(ZipEntry(inMemoryJson.toString()))
+                            Files.copy(inMemoryJson, zipOutputStream)
+                        } else {
+                            throw RuntimeException("Corrupt zip file returned")
+                        }
+                    } else {
+                        zipOutputStream.putNextEntry(ZipEntry(inMemoryProjectPath.toString()))
+                        Files.copy(inMemoryProjectPath, zipOutputStream)
+                    }
+                    zipOutputStream.closeEntry()
+                }
+                zipOutputStream.close()
+                Files.copy(modifiedZipOutputPath, modifiedZipInMemory)
+            }
+
+            if (modifyLocalOnly == false && replaceProject) {
+                // Part II: Upload the modified projects to Inception
+
+                // Step 3: delete the relevant original files.
+                // Since we don't want several "copy_of" projects (which happens
+                // when there is already a project of the same name on Inception),
+                // we delete the original projects before importing the modified ones.
+                logger.info { "Deleting project ${project.key.name}" }
+                retryOnFuelError {
+                    "$inceptionUrl/api/aero/v1/projects/${project.key.id}".httpDelete()
+                            .authenticateToInception()
+                            .retryOnResponseFailure()
+                }
+
+                // Step 4: import the new projects.
+                logger.info { "Importing project $modifiedZipOutputFile" }
+                retryOnFuelError {
+                    // The documentation is a lie.
+                    importUrl.httpUpload()
+                            .add(
+                                    FileDataPart(
+                                            modifiedZipOutputFile,
+                                            name = "file",
+                                            filename = "$modifiedZipOutputFile"
+                                    )
+                            )
+                            .authenticateToInception()
+                            .retryOnResponseFailure()
+                }
+            }
+        } else {
+            failureSet.add(project.key.name)
         }
     }
-    logger.info { "Projects that can safely be deleted: $canBeDeleted" }  // testing
-    modifiedOutputDir.walk().filter { it.isFile }.forEach { zipFile ->
-        val projectName = zipFile.nameWithoutExtension
-        if (canBeDeleted.contains(projectName)) {
-            projectsToDelete.add(projectName)
-            projectsToImport.add(zipFile)
+    logger.info { "All done!" }
+    if (failureSet.size > 0) {
+        logger.info {
+            "The following projects may need their tags added manually: $failureSet"
         }
     }
-    logger.info { "Projects to replace: $projectsToImport" }
-    // Delete projects with replacements
-//    for (project in canBeDeleted) {
-//        logger.info { "Deleting project $project" }
-//        retryOnFuelError {
-//            "$inceptionUrl/api/aero/v1/projects/$project".httpDelete()
-//                    .authenticateToInception()
-//                    .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
-//                    .body
-//        }
-//    }
-//    // Import updated projects to replace deleted ones
-//    for (newProject in projectsToImport) {
-//        logger.info { "Importing project $newProject" }
-//        retryOnFuelError {
-//            "$inceptionUrl/api/aero/v1/projects/import/$newProject".httpPost()
-//                    .authenticateToInception()
-//                    .resultObjectThrowingExceptionOnFailure<AeroResult<Project>>(mapper)
-//                    .body
-//        }
-//    }
+}
+
+private fun addArgumentsToJson(
+        jsonTree: JsonNode, projectName: String, arguments: JsonNode
+): Boolean {
+    /**
+     * Adds the new tags specified for the given project json.
+     *
+     * This also determines whether the project should be replaced
+     * depending on whether any of the new argument labels were
+     * added to the tag list.
+     */
+    val tagSet = (jsonTree["tag_sets"] as ArrayNode)[0] as ObjectNode
+    val tags = tagSet["tags"] as ArrayNode
+
+    // Add arguments
+    val tagObjectMapper = ObjectMapper()
+    val addedArgumentsList = mutableListOf<Boolean>()
+    for (argument in arguments) {
+        val tagJson = tagObjectMapper.createObjectNode()
+        tagJson.set("tag_name", argument)
+        tagJson["tag_description"] = null
+        if (tags.contains(tagJson)) {
+            logger.info { "$projectName already has tag $argument!" }
+            addedArgumentsList.add(false)
+        } else {
+            tags.add(tagJson)
+            addedArgumentsList.add(true)
+        }
+    }
+    logger.info { "Done modifying $projectName's tag set" }
+    return if (addedArgumentsList.all { !it }) {
+        logger.info { "No tags were added to $projectName; it will not be replaced." }
+        false
+    } else { true }
 }

--- a/annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt
+++ b/annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt
@@ -44,7 +44,7 @@ const val EVENT_LOG = "event.log"
  *  </ul>
  */
 
-private data class Project(val id: Long, val name: String)
+data class Project(val id: Long, val name: String)
 class ExportAnnotations {
     companion object {
         fun main(argv: Array<String>) {
@@ -226,7 +226,7 @@ class ExportAnnotations {
     }
 }
 
-private data class Document(val id: Long, val name: String, val state: String) {
+data class Document(val id: Long, val name: String, val state: String) {
     init {
         if (name.isEmpty()) {
             throw IllegalArgumentException("Documents cannot have empty names")
@@ -234,11 +234,11 @@ private data class Document(val id: Long, val name: String, val state: String) {
     }
 }
 
-private data class DocumentResult(val messages: List<String>, val body: List<Document>)
+data class DocumentResult(val messages: List<String>, val body: List<Document>)
 private data class AnnotatorRecord(val user: String, val state: String, val timestamp: String?)
-private data class AeroResult<T>(val messages: List<String>, val body: List<T>)
+data class AeroResult<T>(val messages: List<String>, val body: List<T>)
 
-private inline fun <reified T : Any> Request.resultObjectThrowingExceptionOnFailure(
+inline fun <reified T : Any> Request.resultObjectThrowingExceptionOnFailure(
         mapper: ObjectMapper
 ): T {
     val (_, _, result) = this.responseObject<T>(mapper)

--- a/annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt
+++ b/annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExportAnnotations.kt
@@ -270,7 +270,7 @@ fun <T> retryOnFuelError(maxTries: Int = 3, timeoutInSeconds: Long = 5, function
 }
 
 
-private fun Request.retryOnResponseFailure(maxTries: Int = 3, timeoutInSeconds: Long = 5): ByteArray? {
+fun Request.retryOnResponseFailure(maxTries: Int = 3, timeoutInSeconds: Long = 5): ByteArray? {
     for (i in 1..maxTries) {
         val (_, _, result) = this.response()
         when (result) {

--- a/docs/2020-05-users.txt
+++ b/docs/2020-05-users.txt
@@ -1,0 +1,13 @@
+gabbard
+kirill_kotelnikov
+kyle_valdes
+laura_godlaski
+liz_lee
+lorelei
+natalie
+rodrigo_pecchio
+selene
+sophie_bulman
+tyler_worthy
+wesley_ricketts
+yelena

--- a/docs/aida_changes_0514.json
+++ b/docs/aida_changes_0514.json
@@ -1,0 +1,10 @@
+{
+  "Contact.CommandOrder": ["Topic"],
+  "Contact.CommitmentPromiseExpressIntent": ["Topic"],
+  "Contact.Negotiate": ["Topic"],
+  "Contact.Prevarication": ["Topic"],
+  "Contact.RequestAdvise": ["Topic"],
+  "Contact.ThreatenCoerce": ["Topic"],
+  "Life.Die": ["Killer", "MedicalIssue"],
+  "Life.Injure": ["Injurer", "MedicalIssue"]
+}

--- a/sample_params/add_new_tags.params
+++ b/sample_params/add_new_tags.params
@@ -1,10 +1,11 @@
-INCLUDE generate_projects.common.lee.params
+INCLUDE generate_projects.common.params
 
 inceptionUrl: http://curatedtraining.isi.edu:8080
 inceptionUsername: remote_admin
 # password is expected to be specified in an environmental variable
 inceptionPassword: %INCEPTION_PASSWORD%
 
-project_dir: /Users/isiboston/Documents/projects/gaia/curatedTraining/generated_projects
-modified_output_dir: /Users/isiboston/Documents/projects/gaia/curatedTraining/generated_projects/updated_0514
-new_tag_list: /Users/isiboston/Documents/projects/gaia/curatedTraining/repos/curated-training-annotator/docs/aida_changes_0514.json
+userList: /Users/isiboston/Documents/projects/gaia/curatedTraining/repos/curated-training-annotator/docs/2020-05-users.txt
+originalExportDir: /Users/isiboston/Documents/projects/gaia/curatedTraining/original_exports
+modifiedOutputDir: /Users/isiboston/Documents/projects/gaia/curatedTraining/generated_projects/modified_copies
+newTagList: /Users/isiboston/Documents/projects/gaia/curatedTraining/repos/curated-training-annotator/docs/aida_changes_0514.json

--- a/sample_params/add_new_tags.params
+++ b/sample_params/add_new_tags.params
@@ -1,0 +1,10 @@
+INCLUDE generate_projects.common.lee.params
+
+inceptionUrl: http://curatedtraining.isi.edu:8080
+inceptionUsername: remote_admin
+# password is expected to be specified in an environmental variable
+inceptionPassword: %INCEPTION_PASSWORD%
+
+project_dir: /Users/isiboston/Documents/projects/gaia/curatedTraining/generated_projects
+modified_output_dir: /Users/isiboston/Documents/projects/gaia/curatedTraining/generated_projects/updated_0514
+new_tag_list: /Users/isiboston/Documents/projects/gaia/curatedTraining/repos/curated-training-annotator/docs/aida_changes_0514.json

